src/server/snap.rs

```diff
@@ -239,16 +262,41 @@ impl<R: RaftStoreRouter + 'static> Runnable<Task> for Runner<R> {
                 SNAP_TASK_COUNTER.with_label_values(&["write"]).inc();
                 match self.files.entry(token) {
                     Entry::Occupied(mut e) => {
-                        if let Err(err) = data.write_all_to(&mut e.get_mut().0) {
-                            error!(
-                                "failed to write data to snapshot file {} for token {:?}: {:?}",
-                                e.get_mut().0.path(),
-                                token,
-                                err
-                            );
-                            let (_, msg) = e.remove();
-                            let key = SnapKey::from_snap(msg.get_message().get_snapshot()).unwrap();
-                            self.snap_mgr.deregister(&key, &SnapEntry::Receiving);
+                        let mut finish = true;
+                        {
+                            let (left, right) = data.slice();
+                            if let Err(err) =
+                                write_snap(&mut e.get_mut().0, self.snap_mgr.get_limiter(), left)
+                            {
+                                error!(
+                                    "failed to write data to snapshot file {} for token {:?}: {:?}",
+                                    e.get_mut().0.path(),
+                                    token,
+                                    err
+                                );
+                                let (_, msg) = e.remove();
+                                let key =
+                                    SnapKey::from_snap(msg.get_message().get_snapshot()).unwrap();
+                                self.snap_mgr.deregister(&key, &SnapEntry::Receiving);
+                                finish = false;
+                            } else if let Err(err) =
+                                write_snap(&mut e.get_mut().0, self.snap_mgr.get_limiter(), right)
+                            {
+                                error!(
+                                    "failed to write data to snapshot file {} for token {:?}: {:?}",
+                                    e.get_mut().0.path(),
+                                    token,
+                                    err
+                                );
+                                let (_, msg) = e.remove();
+                                let key =
+                                    SnapKey::from_snap(msg.get_message().get_snapshot()).unwrap();
+                                self.snap_mgr.deregister(&key, &SnapEntry::Receiving);
+                                finish = false;
+                            }
+                        }
+                        if finish {
+                            data.set_empty();
```

Comment:
What's the purpose to set data empty?

Reply:
If the snapshot has been written to the disk successfully, the data should be empty.

Reply:
so any purpose to set empty here?  do we need to check empty later?

Reply:
In original `write`, the data is set empty after finish writing, I think we should leave it that way.